docs/proposals/multipart_trasnfer_mode.md

```diff
@@ -0,0 +1,296 @@
+          "authenticated": true,
+          "header": {
+            "Authorization": "Basic 123abc123abc123abc123abc123="
+          },
+          "expires_in": 86400,
+          "params": {
+            "uploadId": "20492a4d0d84",
+            "partIds": [0, 1, 2, 3]
+          }
+        },
+        "abort": {
+          "href": "https://foo.cloud.com/storage/upload/20492a4d0d84",
+          "authenticated": true,
+          "header": {
+            "Authorization": "Basic 123abc123abc123abc123abc123="
+          },
+          "method": "DELETE",
+          "expires_in": 86400
+        }
+      }
+    }
+  ]
+}
+```
+
+## Uploaded Part Digest
+Some storage backends will support, or even require, uploading clients to send a digest of the uploaded part as part
+of the request. This is a useful capability even if not required, as it allows backends to validate each part
+separately as it is uploaded.
+
+To support this, `parts` request objects may include a `want_digest` value, which may be any value specified by
+[RFC-3230](https://tools.ietf.org/html/rfc3230) or [RFC-5843](https://tools.ietf.org/html/rfc5843) (the design for this
+feature is highly inspired by these RFCs).
+
+Possible values include a comma-separated list of q-factor flagged algorithms, such as SHA-256 and SHA-512.
+Of one or more of these are specified, the digest of the payload is to be specified by the client as part of
+the Digest header, using the format specified by [RFC-3230 section 4.3.2](https://tools.ietf.org/html/rfc3230#section-4.3.1).
+
+It should be noted that digests based on `MD5` or `SHA-1` algorithms are not supported by this transfer mode,
+as they are no longer to be considered secure. Therefore, the `contentMD5`, `MD5` and `SHA` values are not to be used as
+part of `want_digest` headers.
+
+Clients, when receiving a `parts` object with a `want_digest` value, must include in the request to upload the part
+a digest of the part, using the `Digest` HTTP header.
+
+### Uploaded Part Digest Example
+
+#### Examples of a batch response with `want_digest` in the reply
+
+With SHA-512 as a preferred algorithm, and SHA-256 as a less preferred option if SHA-512 is not possible:
+```json
+{
+  "actions": {
+    "parts": [
+      {
+        "href": "https://foo.cloud.com/storage/upload/20492a4d0d84?part=3",
+        "header": {
+          "Authorization": "Bearer someauthorizationtokenwillbesethere"
+        },
+        "pos": 7500001,
+        "want_digest": "sha-512;q=1.0, sha-256;q=0.5"
+      }
+    ]
+  }
+}
+```
+
+#### Example of part upload request send to the storage server
+Following on the `want_digest` value specified in the last example, the client should now send the following headers
+to the server when uploading the part:
+
+```
+HTTP/1.1 PUT /storage/upload/20492a4d0d84?part=3
+Authorization: Bearer someauthorizationtokenwillbesethere
+Digest: SHA-256=thvDyvhfIqlvFe+A9MYgxAfm1q5=,SHA-512=thvDyvhfIqlvFe+A9MYgxAfm1q5thvDyvhfIqlvFe+A9MYgxAfm1q5=
```

Comment:
Can we allow the client to specify just one of these?  I always viewed the negotiation as allowing the client to specify just the most preferred option negotiated.

Reply:
Yes, the client is allowed to do so, as per the `Digest` RFC. Perhaps the example should emphasize that approach instead of sending both.